torch/distributed/distributed_c10d.py

```diff
@@ -1926,7 +1926,7 @@ def broadcast_object_list(object_list, src=0, group=None, device=None):
             if obj_view.device != torch.device("cpu"):
                 obj_view = obj_view.cpu()
             offset += obj_size
-            object_list[i] = _tensor_to_object(obj_view, obj_size)
+            object_list[i] = _tensor_to_object(obj_view, obj_size, device)
```

Comment:
This is semantically different from ZeRO's `_broadcast_object()`. This deserializes the object to the _sender's_ device, not the _receiver's_ device. This can lead to significant performance regression for ZeRO. For example, this means that tensors being broadcast from rank 0 to all ranks will be copied back to rank 0's GPU before inevitably being copied back to the nonzero rank's GPU.

Feel free to refer to this PR for details and a possible approach: https://github.com/pytorch/pytorch/pull/61539
In my understanding, the only deficiency with that PR (causing it to be reverted) is the it did not use the correct pickle protocol.

Reply:
Resolve by re-using the approach from #61539 by separating device and map_location input